src/charm.py

```diff
@@ -71,21 +75,24 @@ def _on_check_runners_action(self, event):
         event.set_results({"status": "Completed runner check"})
 
     def _reconcile_runners(self):
-        delta = self.config["quantity"] - self._runner.active_count()
-        logger.info(f"Reconciling {delta} runners")
-        while delta > 0:
-            self.unit.status = MaintenanceStatus(f"Installing {delta} runner(s)")
-            try:
-                self._runner.create(image="ubuntu", virt=self.config["virt-type"])
-            except RuntimeError as e:
-                logger.error("Failed to create runner")
-                logger.error(f"Error: {e}")
-            delta = self.config["quantity"] - self._runner.active_count()
+        for virt_type in self._virt_type_config_options.keys():
+            target_quantity = self.config[self._virt_type_config_options[virt_type]]
+            delta = target_quantity - self._runner.active_count(virt_type)
+            logger.info(f"Reconciling {delta} runners")
+            while delta > 0:
+                self.unit.status = MaintenanceStatus(f"Installing {delta} runner(s)")
+                try:
+                    self._runner.create(virt_type, "ubuntu")
+                except RuntimeError as e:
+                    logger.error("Failed to create runner")
+                    logger.error(f"Error: {e}")
+                delta = target_quantity - self._runner.active_count(virt_type)
         self.unit.status = ActiveStatus("Active and registered")
 
     def _on_stop(self, event):
         self._remove_cron("check-runners")
-        self._runner.remove_runners()
+        for virt_type in self._virt_type_config_options.keys():
+            self._runner.remove_runners(virt_type)
```

Comment:
Why does the `_virt_type_config_options` dict exist in charms?  it seems linked to the `Runner` class.  If you pushed it into there, you wouldn't have to have these loops in `charms.py`. 

Maybe if this was in `Runner.virt_types` you could iterate on it when you needed it in `charms.py` but leave `remove_runners()` and `active_count` etc down in there. 

It seems like there are operations where you don't care if it's a container or VM, you just want to remove all the runners, or reconcile all the runners.  Dealing with which type isn't specifically important. Here's kinda what i'm thinking

```python
def remove_runners(virt_type=None):
    """
    Removes all the runners of a specific type, or if None specified remove all.
    """
    ...
```

Reply:
The reason for keeping it in the charm.py was to make runner.py virt_type agnostic and to access related config values in one place. I will take a look at your suggestion though.